### PR TITLE
Unite workflow execution api endpoints

### LINKED ISSUE #2899: Reach parity of the old WF controller and the API
When we merged #2077 and #2895 we introduced a technical debt that is caused by non-equivalency of old workflow controller and the workflow API. The GUI now temporarily talks to an inferior internal endpoint and the improvements of such is either limited or will introduce more inconsistency.

We should gather a list of disparities and fix them before 16.10 so we can remove the /api_internal/ route.

@guerler @jmchilton @dannon @blankenberg @natefoo please help me make this list, I took the initial set from @guerler's [comment](https://github.com/galaxyproject/galaxy/pull/2077#issuecomment-223062827)
- [x] Sending the outputs to one or multiple new histories
- [x] Linked batchmode
- [x] Unlinked batchmode
- [x] Support for workflow parameters
- [x] Support for data selection within steps
- [ ] Support for resolving output dataset ids and names
- [ ] Unaddressed queuing effect that the history does not auto-refresh anymore
- [x] Support for multi-selection data inputs (addition @guerler)

And the initial answers from @jmchilton's [comment](https://github.com/galaxyproject/galaxy/pull/2077#issuecomment-223107706) 

> > Support for workflow parameters
> 
> This is supported I believe - https://github.com/galaxyproject/galaxy/blob/dev/test/api/test_workflows.py#L1324.
> 
> > Sending the outputs to one or multiple new histories
> 
> I don't think you should build a WorkflowRunConfig directly - there is an existing abstraction that both API endpoints use and it seems to cover this functionality:
> https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/workflow/run_request.py#L214
> build_workflow_run_config would also cover workflow parameters if used properly.
> 
> > Linked batchmode
> > Unlinked batchmode
> 
> The implementation here is okay and could probably be adapted in the invocation API? I wish it reused the tools stuff and I wish there logic wasn't placed
> 
> > Support for data selection within steps
> 
> If this works - my intuition suggests either you are not properly backgrounding the workflow or encoded IDs are being saved to the database. I can fix the invocation endpoint to accept these - but previous @dannon and I purposely decided that we should discontinue support for non-input data selection. I don't know if his opinions have changed - but mine have't - it is a mistake to allow it. I've outlined how I believe it should be approached - but if you do not wish to do that and insist the API needs to support the data parameters we hoped to deprecate - I will begrudgingly update build_workflow_run_config and the invocation endpoint to allow this (will require a database migration I believe).
> 
> > Support for resolving output dataset ids and names
> > Unaddressed queuing effect that the history does not auto-refresh anymore
> 
> This needs to be resolved longer term (there is an open issue about limitations regarding history updates - fixing it at the history level would help dynamic tool outputs, collection stuff, GIEs that publish to the history, this workflow endpoint, etc...) - but as a basic starting point the invocation landing page could take the invocation ID and poll the workflow and update the history as new datasets are created. Do you want me to do the UI for that?

### PR DESCRIPTION
**Pull request by guerler**
Resolves #2899, branched from #2935.

### CODE REVIEW

**Review by jmchilton — CHANGES_REQUESTED** [review]
I have a couple specific comments also - but at a high-level I was hoping you could throw a flag in there to indicate the response should contain multiple invocations. It seems a bit like magic that if only one invocation if produced a map is returned but if many are produced a list is returned. If the request contained `'multi_request': True` or `batch: True` and then returned a list I would be more comfortable. I'd even more prefer if the batching were done on the client of through the batching stuff Carl wrote - but I understand the pragmatic concerns here.

**Comment by guerler** [comment]
Feel free to correct but if I understand this right, this is already the case. A list of invocations is only returned if one of the parameters has specified `batch: True` otherwise a single map is returned. Besides that, can you help with the test case, it works locally with sqllite and postgres but somehow fails when ran through jenkins. Any ideas? Thanks in advance.

**Comment by guerler** [comment]
The request now always returns a list if `input_params` is specified, https://github.com/galaxyproject/galaxy/pull/2969/commits/1cb67308771edf40716ef5f30f4321051debc4bb#diff-2fca1f4002bba50737993fd85ce5800fR454.

**Comment by jmchilton** [comment]
@guerler I'd really like to unify the `input_params` with the existing `parameters` approach. I don't want the batch mode options to be available in one and not the other - it also isn't very intuitive from the parameter name `input_params` that if it is specified you get a list back. I know it feels redundant in some ways, but I think it would be better if the UI explicitly asked for a list back (`enable_batch_processing: True` or `batch: True`). It would be easier for the endpoint consumer to understand and it would enable us to make more progress toward uniting `input_params` and `parameters`.

**Comment by nsoranzo** [comment]
It would be nice to test if this breaks BioBlend before merging.

**Review by jmchilton — COMMENTED** [review]
Sorry if I seem picky - I do appreciate this particular change - I think it is important improvement.

**Comment by jmchilton** [comment]
@guerler - about this...

```
new_history.copy_tags_from( trans.user, trans.history )
```

is that intentional? That seems wrong - but was it copied from the old code in the UI controller or something?

**Comment by guerler** [comment]
@jmchilton its not a big deal.

**Comment by jmchilton** [comment]
@guerler 

```
                if len( param_keys ) > index:
                    ids = param_keys[ index ]
```

This should just be:

```
                if len( param_keys ) > 0:
                    ids = param_keys[ index ]
```

Or something like that right? param_keys is the hids of things being batched up right?

**Comment by jmchilton** [comment]
I opened some PRs to merge approaches to history handling and make the return value of the run_request method more consistent. Let me know if either is objectionable. I haven't tested them extensively yet but I'd like to know if you are okay with general directions of the refactorings.

**Comment by guerler** [comment]
Alright, I'll take a look. The conditional double checks the result of the parameter expansion. We could remove it entirely, but I do not see any harm leaving it in. The conditional should never fail unless there is a bug in the parameter expansion procedure.

**Comment by jmchilton** [comment]
@guerler I see - that makes sense now - I was mistaking this as the `if nids == 0` case - but  I see that is just taken care of but not doing anything in that nested conditional. I'd drop the new exception and place an assertion in there instead of a conditional - it feels like the most Pythonic thing to do.

Update: This is what I would do (https://github.com/guerler/galaxy/pull/13/commits/ab2a5198afeb0b651643b84c0dd3eb6a0e611bb6) I guess.

**Comment by jmchilton** [comment]
@nsoranzo - I'll definitely run the BioBlend tests and I'll PR enhancements coming out of this work -descriptions of the batch functionality and the flattened parameter style @guerler prefers here.

**Comment by jmchilton** [comment]
@guerler - You can sort keys before iterating over them during expansion to make it deterministic right?

**Comment by jmchilton** [comment]
@guerler I just noticed you are exposing unencoded IDs with the new run export. Can we fix that before the release? I feel strongly that we shouldn't be exposing unencoded IDs over the API - it has long been a design goal of the API I think. It should be possible to represent everything with `order_index` - to whatever extent we can use that the API is more usable and stable to client libraries and humans hacking around.

**Comment by jmchilton** [comment]
https://github.com/guerler/galaxy/pull/13 I think includes a bunch of refactoring so that instead of branching based on approach in `build_workflow_run_config` at the very outer part of a huge function and doing very different things and handling different parameters - everything is handled in one big loop with more consistent parameter handling. The following parameters `history`, `history_id`, `history_id`, `inputs`, `ds_map`, `inputs_by`, `legacy`,  and `no_add_to_history` are now handled consistently across approaches (UI, existing API) - before https://github.com/guerler/galaxy/pull/13 each of these parameters would only apply to one branch or the other of this code. In addition initial `WorkflowRunConfig` objects to be persisted are now only created in one place - one line of code - I think this is strong evidence that we now just have that one path through the code.

The thing that this part of the run UI continues to do which I think is unfortunate is write IDs to JSON blobs in `WorkflowRequestStepState.state` for input steps instead of properly using `WorkflowRequestToInputDatasetAssociation` or `WorkflowRequestToInputDatasetCollectionAssociation` model objects / tables which provide referential integrity and stronger typing since these use the database more appropriately. I don't think that issue needs to hold this up though.

**Review by jmchilton — APPROVED** [review]

File: lib/galaxy/managers/workflows.py
Comment (jmchilton) [inline lib/galaxy/managers/workflows.py]: Do you have a reason not to just use the `order_index` for these? The other export endpoints do this and I think we should be consistent - I also think this makes the API more "usable". I can't think of a reason the GUI should need to know the step ID when the invocation API doesn't use it and the workflow editor doesn't use it - is there one?

Comment (guerler) [reply lib/galaxy/managers/workflows.py]: This id is used to identify connected steps. I used the id because its native to the WorkflowStepConnection model (used below). However, anything which uniquely identifies the steps/connected steps would work.

**Comment by guerler** [comment]
@galaxybot test this

**Comment by guerler** [comment]
@galaxybot test this

**Comment by jmchilton** [comment]
You disappeared from IRC:

```
01:20 <jmchilton> I think you want to sort the keys - not the items 
01:21 <jmchilton> maybe it doesn't matter - but I don't know what sorting a tuple like 
                  that does 
01:22 <jmchilton> I mean - I guess it work on the first item... hmmm... 
01:25 <jmchilton> Okay - I guess the thing is that the invocations do get queued -  
                  there is no guarentee which will start adding to the history first 
01:25 <jmchilton> it is like submitting jobs in a given order and expecting them to  
                  run in the same order - Galaxy doesn't do that 
01:26 <jmchilton> I guess I don't consider that a problem - we could work on FIFO-ing  
                  the workflow scheduler though 
```

**Comment by guerler** [comment]
Alright. Just wanted to make sure that testing both invocation orders is really the right thing to do. Thx.

**Comment by martenson** [comment]
I tried to do some black box testing here and when I try to run a simple wf on this branch it instantly kills my browser tab.

screenshot 2016-10-04 11 05 52

screenshot 2016-10-04 11 06 12

screenshot 2016-10-04 11 05 28

**Comment by martenson** [comment]
update: Chrome with open devtools causes the issue above, with devtools closed it runs fine

Another issue is that if you open the run workflow in incognito window it will fail to render properly

screenshot 2016-10-04 12 23 29

**Comment by guerler** [comment]
Very good catch. Hiding the child forms instead of removing them avoids the crash when Chromes devtools are open. I could not reproduce the second issue yet.

**Comment by martenson** [comment]
I can produce the incognito issue reliably on a WF that has only the 'merge columns' tool in it.

**Comment by martenson** [comment]
as a sideeffect of 329ee052510c0c720ae5d44bac23becf639f6506 the submission is _way_ faster

**Comment by martenson** [comment]
the second issue was caused by a broken WF, when exported it has no steps:

``` json
{
    "a_galaxy_workflow": "true", 
    "annotation": "", 
    "format-version": "0.1", 
    "name": "merge test", 
    "steps": {}, 
    "uuid": "87df19a4-454d-445b-ba2b-0f35676e14a0"
}
```

I think we should make WF creation atomic operation, but that is not in scope of this PR.

**Review by martenson — APPROVED** [review]
Looks and works very fine to me. Thank you duo of  @guerler & @jmchilton

**Comment by jmchilton** [comment]
This branch seems to cause a bioblend test failure (most though not 100% of the time).

Against dev:

```
BIOBLEND_GALAXY_API_KEY=a7bf49fe4fd4041a2a93de4d0901716b BIOBLEND_GALAXY_URL=http://localhost:8080/ nosetests tests/TestGalaxyWorkflows.py
test_cancelling_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_export_workflow_to_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_get_workflows (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_import_workflow_from_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_run_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_show_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok

----------------------------------------------------------------------
Ran 7 tests in 15.312s

OK
```

Against this branch:

```
BIOBLEND_GALAXY_API_KEY=a7bf49fe4fd4041a2a93de4d0901716b BIOBLEND_GALAXY_URL=http://localhost:8080/ nosetests tests/TestGalaxyWorkflows.py
test_cancelling_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_export_workflow_to_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_get_workflows (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_import_workflow_from_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_run_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_show_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... FAIL

======================================================================
FAIL: test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/TestGalaxyWorkflows.py", line 54, in test_workflow_scheduling
    self.assertEqual(invocation['state'], "ready")
AssertionError: u'new' != 'ready'
-------------------- >> begin captured logging << --------------------
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/upload?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/histories?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/tools?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/59c76a119581e190/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/59c76a119581e190/invocations/e89067bb68bee7a0?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 7 tests in 25.187s

FAILED (failures=1)
```

After dropping that assertion and letting the test continue:

```
BIOBLEND_GALAXY_API_KEY=a7bf49fe4fd4041a2a93de4d0901716b BIOBLEND_GALAXY_URL=http://localhost:8080/ nosetests tests/TestGalaxyWorkflows.py
test_cancelling_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_export_workflow_to_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_get_workflows (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_import_workflow_from_local_path (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_run_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_show_workflow (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ok
test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows) ... ERROR

======================================================================
ERROR: test_workflow_scheduling (tests.TestGalaxyWorkflows.TestGalaxyWorkflows)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/test_util.py", line 54, in wrapped_method
    return method(has_gi, *args, **kwargs)
  File "/home/john/workspace/bioblend/tests/TestGalaxyWorkflows.py", line 57, in test_workflow_scheduling
    pause_step = steps[2]
KeyError: 2
-------------------- >> begin captured logging << --------------------
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/tools?in_panel=False&key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/upload?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/histories?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/tools?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "POST /api/workflows/5a9381f02a7b89af/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
requests.packages.urllib3.connectionpool: INFO: Starting new HTTP connection (1): localhost
requests.packages.urllib3.connectionpool: DEBUG: "GET /api/workflows/5a9381f02a7b89af/invocations/a7db2fac67043c7e?key=a7bf49fe4fd4041a2a93de4d0901716b HTTP/1.1" 200 None
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 7 tests in 22.767s

FAILED (errors=1)
```

The failing test https://github.com/galaxyproject/bioblend/blob/master/tests/TestGalaxyWorkflows.py#L25. I wrote the test and am quite surprised it is testing something not tested by Galaxy itself - I'll look into this some more.

**Comment by jmchilton** [comment]
Appears to be something up with the history id being empty on the resulting invocation.

**Pull request merged**
